File: lua-lib/aiomsg/socket.lua
-- aiomsg — native Lua smart sockets (a cooperative LuaSocket reactor).
--
-- A single Socket multiplexes many TCP connections behind one object, with
-- ZMQ-like distribution patterns (publish / round-robin / by-identity),
-- automatic reconnection, send buffering, heartbeating, an optional
-- at-least-once delivery guarantee, and TLS (LuaSec). It speaks the
-- language-independent aiomsg wire protocol (see ../PROTOCOL.md) and
-- interoperates on the wire with the Python reference and every other port.
--
-- Concurrency model. Lua has no threads, so — like every LuaSocket program of
-- any size — the socket owns a single cooperative reactor built on non-blocking
-- sockets and socket.select. There is no background execution: the reactor is
-- driven by the calls that wait on it. recv() pumps it until a message arrives;
-- run(seconds) pumps it for a fixed time (to keep connections alive, flush
-- sends, answer heartbeats and acks) without receiving. All connection
-- management — handshakes, heartbeats, dead-peer detection, reconnection,
-- at-least-once resends — happens inside that one loop.

local socket = require("socket")
local protocol = require("aiomsg.protocol")
local tls = require("aiomsg.tls")

local now = socket.gettime

math.randomseed(math.floor((now() * 1e6) % 2147483647))

local HEARTBEAT = 5.0
local TIMEOUT = 15.0
local HANDSHAKE_TIMEOUT = 15.0
local CONNECT_TIMEOUT = 1.0
local RECONNECT = 0.1
local RESEND = 5.0
local MAX_RETRIES = 5
local READ_CHUNK = 65536
local MAX_IDLE_TICK = 1.0 -- cap select() blocking so timers stay responsive

local M = {}

M.SendMode = { ROUND_ROBIN = "round-robin", PUBLISH = "publish" }
M.Delivery = { AT_MOST_ONCE = "at-most-once", AT_LEAST_ONCE = "at-least-once" }

local function random_bytes(n)
  local chars = {}
  for i = 1, n do
    chars[i] = string.char(math.random(0, 255))
  end
  return table.concat(chars)
end

local function to_hex(bytes)
  return (bytes:gsub(".", function(c)
    return string.format("%02x", string.byte(c))
  end))
end

local Socket = {}
Socket.__index = Socket
M.Socket = Socket

-- opts = {send_mode=, delivery=, identity=}
function Socket.new(opts)
  opts = opts or {}
  local self = setmetatable({}, Socket)
  self.send_mode = opts.send_mode or M.SendMode.ROUND_ROBIN
  self.delivery = opts.delivery or M.Delivery.AT_MOST_ONCE
  self.id = opts.identity or random_bytes(protocol.IDENTITY_SIZE)

  self.conns = {} -- active (post-handshake) connections, in round-robin order
  self.rr = 0
  self.buffer = {} -- {target=, data=} queued while no peer is connected
  self.pending = {} -- hex(msg_id) -> {target=, data=, retries=, deadline=}

  self.servers = {} -- listening sockets -> tls params or false
  self.connectors = {} -- {host=, port=, tls=, conn=, retry_at=}
  self.all_conns = {} -- every connection object, active or handshaking
  self.inbox = {} -- {data=, sender=} delivered to the application
  self.closed = false
  return self
end

function Socket:identity()
  return self.id
end

-- --- bind / connect ----------------------------------------------------

-- tls_params (optional): {cert=path, key=path} to listen for TLS peers.
function Socket:bind(host, port, tls_params)
  local server, err = socket.bind(host, port)
  if not server then
    error("bind failed: " .. tostring(err))
  end
  server:settimeout(0)
  self.servers[server] = tls_params and tls.server_params(tls_params.cert, tls_params.key) or false
  return self
end

-- tls_params (optional): {ca=path, server_name=name} to connect over TLS.
function Socket:connect(host, port, tls_params)
  local params = nil
  if tls_params then
    local name = tls_params.server_name
    if name == nil or name == "" then
      name = host
    end
    params = tls.client_params(tls_params.ca, name)
  end
  self.connectors[#self.connectors + 1] =
    { host = host, port = port, tls = params, conn = nil, retry_at = now() }
  return self
end

-- --- send / receive ----------------------------------------------------

function Socket:send(data)
  self:_dispatch(nil, data)
  self:_step(0) -- make progress without blocking
end

function Socket:send_to(identity, data)
  self:_dispatch(identity, data)
  self:_step(0)
end

-- Block (pumping the reactor) until the next message; returns data, sender.
-- With a timeout (seconds) returns nil on expiry. Returns nil once closed.
function Socket:recv(timeout)
  local deadline = timeout and (now() + timeout) or nil
  while true do
    if #self.inbox > 0 then
      local message = table.remove(self.inbox, 1)
      return message.data, message.sender
    end
    if self.closed then
      return nil
    end
    local tick = self:_next_timeout()
    if deadline then
      local remaining = deadline - now()
      if remaining <= 0 then
        return nil
      end
      tick = math.min(tick, remaining)
    end
    self:_step(tick)
  end
end

-- Iterator over incoming messages: `for data, sender in sock:messages() do`.
function Socket:messages()
  return function()
    return self:recv()
  end
end

-- Pump the reactor for `seconds` (or forever if nil) without receiving. Use to
-- keep connections alive long enough to flush sends and answer acks.
function Socket:run(seconds)
  local deadline = seconds and (now() + seconds) or nil
  while not self.closed do
    local tick = self:_next_timeout()
    if deadline then
      local remaining = deadline - now()
      if remaining <= 0 then
        return
      end
      tick = math.min(tick, remaining)
    end
    self:_step(tick)
  end
end

-- Advance the reactor by a single turn, blocking at most `timeout` seconds (0 =
-- non-blocking). Useful for driving more than one socket from one thread; recv()
-- and run() are the usual single-socket drivers.
function Socket:poll(timeout)
  self:_step(timeout or 0)
end

function Socket:close()
  if self.closed then
    return
  end
  self.closed = true
  for server in pairs(self.servers) do
    server:close()
  end
  for _, conn in ipairs(self.all_conns) do
    pcall(function()
      conn.sock:close()
    end)
  end
end

-- --- broker (routing, buffering, delivery, at-least-once) --------------

local function find_conn(self, identity)
  for _, conn in ipairs(self.conns) do
    if conn.peer == identity then
      return conn
    end
  end
  return nil
end

local function queue_frame(conn, framed)
  conn.outbuf = conn.outbuf .. framed
  conn.last_send = now() -- any real frame resets the heartbeat interval
end

function Socket:_route(target, framed)
  if self.closed or #self.conns == 0 then
    return
  end
  if target then
    local conn = find_conn(self, target) -- by-identity: drop if peer is gone
    if conn then
      queue_frame(conn, framed)
    end
  elseif self.send_mode == M.SendMode.PUBLISH then
    for _, conn in ipairs(self.conns) do
      queue_frame(conn, framed)
    end
  else
    self.rr = self.rr % #self.conns + 1
    queue_frame(self.conns[self.rr], framed)
  end
end

function Socket:_transmit(target, data, retries)
  local at_least_once = self.delivery == M.Delivery.AT_LEAST_ONCE
    and (target ~= nil or self.send_mode == M.SendMode.ROUND_ROBIN)
  if not at_least_once then
    self:_route(target, protocol.frame_data(data))
    return
  end
  local msg_id = random_bytes(protocol.MSG_ID_SIZE)
  self.pending[to_hex(msg_id)] =
    { target = target, data = data, retries = retries, deadline = now() + RESEND }
  self:_route(target, protocol.frame_data_req(msg_id, data))
end

function Socket:_dispatch(target, data)
  if self.closed then
    return
  end
  if #self.conns == 0 then
    self.buffer[#self.buffer + 1] = { target = target, data = data }
  else
    self:_transmit(target, data, MAX_RETRIES)
  end
end

function Socket:_flush_buffer()
  local queued = self.buffer
  self.buffer = {}
  for _, item in ipairs(queued) do
    self:_dispatch(item.target, item.data)
  end
end

function Socket:_received(sender, envelope)
  if envelope.type == protocol.TYPE.DATA then
    self.inbox[#self.inbox + 1] = { data = envelope.payload, sender = sender }
  elseif envelope.type == protocol.TYPE.DATA_REQ then
    self.inbox[#self.inbox + 1] = { data = envelope.payload, sender = sender }
    self:_route(sender, protocol.frame_ack(envelope.msg_id)) -- ack on same conn
  elseif envelope.type == protocol.TYPE.ACK then
    self.pending[to_hex(envelope.msg_id)] = nil
  end
end

function Socket:_sweep()
  local current = now()
  for key, p in pairs(self.pending) do
    if p.deadline <= current then
      self.pending[key] = nil
      if p.retries > 0 then
        if #self.conns == 0 then
          self.buffer[#self.buffer + 1] = { target = p.target, data = p.data }
        else
          self:_transmit(p.target, p.data, p.retries - 1)
        end
      end
    end
  end
end

-- --- connection lifecycle ---------------------------------------------

-- Create a connection object wrapping a freshly-accepted or freshly-connected
-- non-blocking socket. `connector` is set for client connections (for reconnect).
local function new_conn(sock, is_server, connector)
  return {
    sock = sock,
    is_server = is_server,
    connector = connector,
    state = "connecting", -- connecting -> tls -> open
    peer = nil,
    handshaked = false,
    decoder = protocol.Decoder.new(),
    outbuf = "",
    last_send = now(),
    last_recv = now(),
    deadline = now() + CONNECT_TIMEOUT,
  }
end

function Socket:_start_connect(connector)
  local sock = socket.tcp()
  sock:settimeout(0)
  local conn = new_conn(sock, false, connector)
  connector.conn = conn
  self.all_conns[#self.all_conns + 1] = conn
  -- Non-blocking connect returns immediately; completion shows up as writable.
  sock:connect(connector.host, connector.port)
end

function Socket:_accept(server, tls_params)
  local client = server:accept()
  if not client then
    return
  end
  client:settimeout(0)
  local conn = new_conn(client, true, nil)
  self.all_conns[#self.all_conns + 1] = conn
  if tls_params then
    conn.tls = tls_params
    conn.sock = tls.wrap(client, tls_params)
    conn.sock:settimeout(0)
    conn.state = "tls"
  else
    conn.state = "open"
    queue_frame(conn, protocol.frame_hello(self.id)) -- our half of the handshake
  end
end

-- A client TCP connect has completed (socket is writable): verify it, then
-- either start the TLS handshake or send HELLO.
function Socket:_connected(conn)
  if not conn.sock:getpeername() then
    self:_drop(conn) -- connect failed
    return
  end
  if conn.connector.tls then
    conn.tls = conn.connector.tls
    conn.sock = tls.wrap(conn.sock, conn.tls)
    conn.sock:settimeout(0)
    conn.state = "tls"
    conn.deadline = now() + HANDSHAKE_TIMEOUT
  else
    conn.state = "open"
    conn.deadline = now() + HANDSHAKE_TIMEOUT
    queue_frame(conn, protocol.frame_hello(self.id))
  end
end

-- Drive a non-blocking TLS handshake one step; returns when it can't progress.
function Socket:_tls_step(conn)
  local ok, err = conn.sock:dohandshake()
  if ok then
    conn.state = "open"
    conn.deadline = now() + HANDSHAKE_TIMEOUT
    queue_frame(conn, protocol.frame_hello(self.id))
  elseif err ~= "wantread" and err ~= "wantwrite" and err ~= "timeout" then
    self:_drop(conn) -- handshake failed
  end
end

-- Read whatever bytes are available on a non-blocking socket. Returns the bytes
-- (possibly "") and whether the peer closed the connection.
local function read_available(sock)
  local parts = {}
  while true do
    local chunk, err, partial = sock:receive(READ_CHUNK)
    if chunk then
      parts[#parts + 1] = chunk
    else
      if partial and #partial > 0 then
        parts[#parts + 1] = partial
      end
      if err == "closed" then
        return table.concat(parts), true
      end
      return table.concat(parts), false -- timeout / wantread: nothing more now
    end
  end
end

function Socket:_on_readable(conn)
  local data, peer_closed = read_available(conn.sock)
  if #data > 0 then
    conn.last_recv = now()
    conn.decoder:push(data)
    while true do
      local envelope = conn.decoder:pop()
      if not envelope then
        break
      end
      local parsed = protocol.parse_envelope(envelope)
      if parsed then
        if not conn.handshaked then
          if not self:_complete_handshake(conn, parsed) then
            return
          end
        elseif parsed.type ~= protocol.TYPE.HEARTBEAT then
          self:_received(conn.peer, parsed)
        end
      end
    end
  end
  if peer_closed then
    self:_drop(conn)
  end
end

-- Validate the peer's HELLO and register the connection, or drop it.
function Socket:_complete_handshake(conn, envelope)
  if envelope.type ~= protocol.TYPE.HELLO or envelope.version ~= protocol.VERSION then
    self:_drop(conn)
    return false
  end
  if find_conn(self, envelope.identity) then
    self:_drop(conn) -- duplicate identity: keep the existing connection
    return false
  end
  conn.peer = envelope.identity
  conn.handshaked = true
  self.conns[#self.conns + 1] = conn
  self:_flush_buffer()
  return true
end

-- Flush as much of conn.outbuf as the socket will take right now. LuaSocket's
-- send returns the index of the last byte sent on success, or nil,err,last on a
-- partial write; either way we trim exactly what went out and keep the rest for
-- the next writable event.
function Socket:_on_writable(conn)
  if #conn.outbuf == 0 then
    return
  end
  local i, err, last = conn.sock:send(conn.outbuf)
  local sent = i or last or 0
  conn.outbuf = string.sub(conn.outbuf, sent + 1)
  if err == "closed" then
    self:_drop(conn)
  end
end

local function remove_from(list, item)
  for i, value in ipairs(list) do
    if value == item then
      table.remove(list, i)
      return
    end
  end
end

function Socket:_drop(conn)
  if conn.dropped then
    return
  end
  conn.dropped = true
  pcall(function()
    conn.sock:close()
  end)
  remove_from(self.conns, conn)
  remove_from(self.all_conns, conn)
  if conn.connector then
    conn.connector.conn = nil
    conn.connector.retry_at = now() + RECONNECT
  end
end

-- Smallest delay until something needs doing (heartbeat, timeout, resend,
-- reconnect), so select() never sleeps past a due timer.
function Socket:_next_timeout()
  local tick = MAX_IDLE_TICK
  local current = now()
  for _, conn in ipairs(self.conns) do
    tick = math.min(tick, (conn.last_send + HEARTBEAT) - current)
    tick = math.min(tick, (conn.last_recv + TIMEOUT) - current)
  end
  for _, conn in ipairs(self.all_conns) do
    if conn.state ~= "open" or not conn.handshaked then
      tick = math.min(tick, conn.deadline - current)
    end
  end
  for _, connector in ipairs(self.connectors) do
    if not connector.conn then
      tick = math.min(tick, connector.retry_at - current)
    end
  end
  if next(self.pending) then
    tick = math.min(tick, RESEND)
  end
  return math.max(tick, 0)
end

-- A shallow copy, so handlers can drop connections while we iterate.
local function copy_list(list)
  local out = {}
  for i, value in ipairs(list) do
    out[i] = value
  end
  return out
end

-- One turn of the reactor: start due (re)connects, select on every socket for
-- up to `timeout` seconds, service whatever is ready, then run the timers.
function Socket:_step(timeout)
  local current = now()

  -- Start any pending (re)connections whose backoff has elapsed.
  if not self.closed then
    for _, connector in ipairs(self.connectors) do
      if not connector.conn and current >= connector.retry_at then
        self:_start_connect(connector)
      end
    end
  end

  -- Build the select sets and a reverse map from socket back to its connection.
  local readers, writers, conn_by_sock = {}, {}, {}
  for server in pairs(self.servers) do
    readers[#readers + 1] = server
  end
  for _, conn in ipairs(self.all_conns) do
    conn_by_sock[conn.sock] = conn
    if conn.state == "connecting" then
      writers[#writers + 1] = conn.sock
    elseif conn.state == "tls" then
      readers[#readers + 1] = conn.sock
      writers[#writers + 1] = conn.sock
    else -- open
      readers[#readers + 1] = conn.sock
      if #conn.outbuf > 0 then
        writers[#writers + 1] = conn.sock
      end
    end
  end

  if #readers == 0 and #writers == 0 then
    if timeout > 0 then
      socket.sleep(timeout)
    end
  else
    local readable, writable = socket.select(readers, writers, timeout)
    for _, sock in ipairs(writable) do
      local conn = conn_by_sock[sock]
      if conn and not conn.dropped then
        if conn.state == "connecting" then
          self:_connected(conn)
        elseif conn.state == "tls" then
          self:_tls_step(conn)
        end
        if conn.state == "open" and not conn.dropped then
          self:_on_writable(conn)
        end
      end
    end
    for _, sock in ipairs(readable) do
      local conn = conn_by_sock[sock]
      if conn then
        if not conn.dropped and conn.state == "tls" then
          self:_tls_step(conn)
        end
        if not conn.dropped and conn.state == "open" then
          self:_on_readable(conn)
        end
      else
        self:_accept(sock, self.servers[sock]) -- a listening server
      end
    end
  end

  self:_process_timers()
end

-- Heartbeats on send-idle connections, dead-peer detection on receive-idle ones,
-- handshake/connect deadlines, and at-least-once resends.
function Socket:_process_timers()
  local current = now()
  for _, conn in ipairs(copy_list(self.all_conns)) do
    if (conn.state ~= "open" or not conn.handshaked) and current > conn.deadline then
      self:_drop(conn)
    end
  end
  for _, conn in ipairs(copy_list(self.conns)) do
    if current - conn.last_recv >= TIMEOUT then
      self:_drop(conn)
    elseif current - conn.last_send >= HEARTBEAT then
      queue_frame(conn, protocol.frame_heartbeat())
    end
  end
  self:_sweep()
end

return M

File: javascript-lib/src/socket.js
// aiomsg — native JavaScript smart sockets (Node's async, event-driven I/O).
//
// A single Socket multiplexes many TCP connections behind one object, with
// ZMQ-like distribution patterns (publish / round-robin / by-identity),
// automatic reconnection, send buffering, heartbeating, an optional
// at-least-once delivery guarantee, and TLS. It speaks the language-independent
// aiomsg wire protocol (see ../PROTOCOL.md) and interoperates on the wire with
// the Python reference and every other port.
//
// Concurrency model. Node is single-threaded around a non-blocking event loop,
// so there is no shared-state locking to do: broker state (the connection list,
// round-robin cursor, send buffer, in-flight ack table) is only ever touched
// from loop callbacks, which never run concurrently. Each connection is a
// net/tls socket with its own heartbeat timer (fires after send-inactivity) and
// receive-timeout timer (resets on every inbound chunk). Writing to an
// otherwise-idle peer is immediate — there is no polling latency.

import net from "node:net";
import tls from "node:tls";
import { randomBytes } from "node:crypto";

import {
  Decoder,
  IDENTITY_SIZE,
  MSG_ID_SIZE,
  PROTOCOL_VERSION,
  Type,
  frameAck,
  frameData,
  frameDataReq,
  frameHeartbeat,
  frameHello,
  parseEnvelope,
} from "./protocol.js";
import { sniffAndUpgrade } from "./ws.js";

export const SendMode = Object.freeze({ ROUND_ROBIN: "round-robin", PUBLISH: "publish" });
export const Delivery = Object.freeze({ AT_MOST_ONCE: "at-most-once", AT_LEAST_ONCE: "at-least-once" });

const HEARTBEAT_MS = 5000;
const TIMEOUT_MS = 15000;
const RECONNECT_MS = 100;
const CONNECT_TIMEOUT_MS = 1000;
const RESEND_MS = 5000;
const MAX_RETRIES = 5;

/**
 * An async, unbounded queue with a single semantic: take() resolves with the
 * next item, or with null once the queue is closed and drained. This is what
 * turns the callback-driven receive path into the awaitable recv()/messages()
 * API without exposing any of the plumbing.
 */
class Inbox {
  #items = [];
  #waiters = [];
  #closed = false;

  push(item) {
    const waiter = this.#waiters.shift();
    if (waiter) waiter(item);
    else this.#items.push(item);
  }

  take() {
    if (this.#items.length) return Promise.resolve(this.#items.shift());
    if (this.#closed) return Promise.resolve(null);
    return new Promise((resolve) => this.#waiters.push(resolve));
  }

  close() {
    this.#closed = true;
    for (const waiter of this.#waiters) waiter(null);
    this.#waiters = [];
  }
}

/** State for one live connection: its socket, peer identity, frame decoder,
 * and the two timers that implement heartbeating and dead-peer detection. */
class Conn {
  constructor(socket) {
    this.socket = socket;
    this.peer = null; // 16-byte identity, set at handshake
    this.decoder = new Decoder();
    this.handshaked = false;
    this.heartbeatTimer = null;
    this.timeoutTimer = null;
  }
}

/** A received message: the payload plus the identity of the peer it came from. */
export class Message {
  constructor(data, sender) {
    this.data = data;
    this.sender = sender;
  }
}

export class Socket {
  /**
   * @param {object} [opts]
   * @param {string} [opts.sendMode] one of SendMode.* (default round-robin)
   * @param {string} [opts.delivery] one of Delivery.* (default at-most-once)
   * @param {Buffer} [opts.identity] 16 raw bytes; a random one is generated otherwise
   */
  constructor({ sendMode = SendMode.ROUND_ROBIN, delivery = Delivery.AT_MOST_ONCE, identity } = {}) {
    this.sendMode = sendMode;
    this.delivery = delivery;
    this.id = identity ? Buffer.from(identity) : randomBytes(IDENTITY_SIZE);

    this.conns = []; // registered (post-handshake) connections
    this.rrCursor = 0;
    this.buffer = []; // {target, data} queued while no peer is connected
    this.pending = new Map(); // msgIdHex -> {timer, target, data, retries}

    this.inbox = new Inbox();
    this.servers = new Set();
    this.sockets = new Set(); // every socket we own, for teardown
    this.reconnectTimers = new Set(); // pending reconnect timers, for teardown
    this.closed = false;
  }

  /** This socket's 16-byte identity (a copy). */
  identity() {
    return Buffer.from(this.id);
  }

  // --- bind / connect ----------------------------------------------------

  /**
   * Listen for peers. Resolves once the listening socket is up.
   * @param {string} host
   * @param {number} port
   * @param {object} [opts] {tls: {key, cert}} for a TLS server
   */
  bind(host, port, { tls: tlsOpts } = {}) {
    return new Promise((resolve, reject) => {
      const onConn = (socket) => this.#accept(socket);
      const server = tlsOpts
        ? tls.createServer({ ...tlsOpts }, onConn)
        : net.createServer(onConn);
      server.on("error", reject);
      this.servers.add(server);
      server.listen(port, host, () => {
        server.off("error", reject);
        server.on("error", () => {}); // ignore post-listen errors
        resolve();
      });
    });
  }

  /**
   * Connect to a peer, reconnecting for the life of the socket. Returns once
   * the reconnect loop has started (the connection itself is established
   * asynchronously, exactly as a real network connect would be).
   * @param {string} host
   * @param {number} port
   * @param {object} [opts] {tls: {ca, servername}} for a TLS client
   */
  connect(host, port, { tls: tlsOpts } = {}) {
    this.#connectLoop(host, port, tlsOpts);
    return Promise.resolve();
  }

  // --- send / receive ----------------------------------------------------

  /** Send to peers per the send mode. Buffered if no peer is connected yet. */
  send(data) {
    this.#dispatch(null, Buffer.from(data));
  }

  /** Send directly to the peer with the given identity, regardless of mode. */
  sendTo(identity, data) {
    this.#dispatch(Buffer.from(identity), Buffer.from(data));
  }

  /** The next message's payload, or null once the socket is closed and drained. */
  async recv() {
    const msg = await this.inbox.take();
    return msg ? msg.data : null;
  }

  /** The next message as a {@link Message} (payload + sender), or null at EOF. */
  recvMessage() {
    return this.inbox.take();
  }

  /** Async iterator over message payloads, ending when the socket closes. */
  async *messages() {
    for (let msg; (msg = await this.inbox.take()) !== null; ) yield msg.data;
  }

  /** Shut everything down: servers, connections, timers, and the inbox. */
  close() {
    if (this.closed) return;
    this.closed = true;
    for (const { timer } of this.pending.values()) clearTimeout(timer);
    this.pending.clear();
    for (const timer of this.reconnectTimers) clearTimeout(timer);
    this.reconnectTimers.clear();
    for (const server of this.servers) server.close();
    for (const socket of this.sockets) socket.destroy();
    this.inbox.close();
  }

  // --- connection lifecycle ---------------------------------------------

  #connectLoop(host, port, tlsOpts) {
    if (this.closed) return;
    // The retry timer stays ref'd so a connect-mode socket keeps the process
    // alive between attempts (an unref'd timer lets the event loop drain and
    // the process silently exit while the peer is still coming up); close()
    // cancels it via reconnectTimers.
    const retry = () => {
      if (this.closed) return;
      const timer = setTimeout(() => {
        this.reconnectTimers.delete(timer);
        this.#connectLoop(host, port, tlsOpts);
      }, RECONNECT_MS);
      this.reconnectTimers.add(timer);
    };

    const socket = tlsOpts
      ? tls.connect({ host, port, ...tlsOpts, timeout: CONNECT_TIMEOUT_MS })
      : net.connect({ host, port, timeout: CONNECT_TIMEOUT_MS });
    // Connect-timeout only guards the initial connect; clear it once up.
    socket.once(tlsOpts ? "secureConnect" : "connect", () => socket.setTimeout(0));
    socket.once("error", () => {});
    // Whatever ends this connection (failure or later drop), reconnect.
    socket.once("close", retry);
    this.#runConnection(socket, false);
  }

  // Bind-side accept: sniff raw-vs-WebSocket (PROTOCOL.md §10) before running
  // the ordinary connection handler. The connect end never receives WebSocket,
  // so it calls #runConnection directly.
  async #accept(socket) {
    const result = await sniffAndUpgrade(socket);
    if (!result || this.closed) {
      if (result) result.socket.destroy();
      return;
    }
    this.#runConnection(result.socket, true, result.prebuffer);
  }

  // Drive one connection: HELLO exchange, registration, then frame delivery
  // until the peer goes quiet or away. `prebuffer` carries bytes already read
  // during the raw-vs-WS sniff (bind path only) that must seed the decoder.
  #runConnection(socket, isServer, prebuffer) {
    this.sockets.add(socket);
    const conn = new Conn(socket);
    socket.setNoDelay(true);

    const start = () => {
      this.#write(conn, frameHello(this.id)); // our half of the symmetric handshake
      this.#resetTimeout(conn);
    };
    if (isServer || !(socket instanceof tls.TLSSocket)) start();
    else socket.once("secureConnect", start);

    socket.on("data", (chunk) => this.#onData(conn, chunk));
    socket.on("error", () => {}); // 'close' handles teardown
    socket.once("close", () => this.#teardown(conn));

    // Seed the decoder with bytes consumed by the sniff before this handler's
    // 'data' listener existed (the sniffed first byte, and any pipelined frames).
    if (prebuffer && prebuffer.length) this.#onData(conn, prebuffer);
  }

  #onData(conn, chunk) {
    this.#resetTimeout(conn);
    conn.decoder.push(chunk);
    for (let env; (env = conn.decoder.pop()) !== null; ) {
      const e = parseEnvelope(env);
      if (e === null) continue; // unknown / truncated type: skip per protocol
      if (!conn.handshaked) {
        if (!this.#completeHandshake(conn, e)) return; // bad HELLO: connection dropped
        continue;
      }
      this.#received(conn.peer, e);
    }
  }

  // Validate the peer's HELLO and register the connection, or drop it.
  #completeHandshake(conn, e) {
    if (e.type !== Type.HELLO || e.version !== PROTOCOL_VERSION) {
      conn.socket.destroy();
      return false;
    }
    if (this.#findConn(e.identity)) {
      conn.socket.destroy(); // duplicate identity: keep the existing connection
      return false;
    }
    conn.peer = e.identity;
    conn.handshaked = true;
    this.conns.push(conn);
    this.#flushBuffer();
    return true;
  }

  #teardown(conn) {
    this.sockets.delete(conn.socket);
    clearTimeout(conn.heartbeatTimer);
    clearTimeout(conn.timeoutTimer);
    const i = this.conns.indexOf(conn);
    if (i >= 0) this.conns.splice(i, 1);
  }

  // --- timers ------------------------------------------------------------

  // Send a heartbeat after HEARTBEAT_MS of send-inactivity; any real send
  // resets this (see #write), so heartbeats only fill genuine idle gaps.
  #scheduleHeartbeat(conn) {
    clearTimeout(conn.heartbeatTimer);
    conn.heartbeatTimer = setTimeout(() => {
      if (!conn.socket.destroyed) this.#write(conn, frameHeartbeat());
    }, HEARTBEAT_MS);
    conn.heartbeatTimer.unref();
  }

  // Tear the connection down if no frame of any type arrives within TIMEOUT_MS.
  #resetTimeout(conn) {
    clearTimeout(conn.timeoutTimer);
    conn.timeoutTimer = setTimeout(() => conn.socket.destroy(), TIMEOUT_MS);
    conn.timeoutTimer.unref();
  }

  #write(conn, frame) {
    conn.socket.write(frame);
    this.#scheduleHeartbeat(conn);
  }

  // --- broker (routing, buffering, delivery, at-least-once) --------------

  #dispatch(target, data) {
    if (this.closed) return;
    if (this.conns.length === 0) this.buffer.push({ target, data });
    else this.#transmit(target, data, MAX_RETRIES);
  }

  #findConn(identity) {
    return this.conns.find((c) => c.peer.equals(identity));
  }

  #route(target, frame) {
    if (this.closed || this.conns.length === 0) return;
    if (target) {
      const c = this.#findConn(target);
      if (c) this.#write(c, frame); // by-identity: drop if that peer is gone
    } else if (this.sendMode === SendMode.PUBLISH) {
      for (const c of this.conns) this.#write(c, frame);
    } else {
      this.#write(this.conns[this.rrCursor++ % this.conns.length], frame);
    }
  }

  // Send one message, choosing DATA vs DATA_REQ by the delivery guarantee.
  // AT_LEAST_ONCE applies only to by-identity and round-robin sends (never
  // PUBLISH); for those a fresh MSG_ID is tracked and resent on timeout.
  #transmit(target, data, retries) {
    const atLeastOnce =
      this.delivery === Delivery.AT_LEAST_ONCE && (target || this.sendMode === SendMode.ROUND_ROBIN);
    if (!atLeastOnce) {
      this.#route(target, frameData(data));
      return;
    }
    const msgId = randomBytes(MSG_ID_SIZE);
    const key = msgId.toString("hex");
    const timer = setTimeout(() => this.#resend(key), RESEND_MS);
    timer.unref();
    this.pending.set(key, { timer, target, data, retries });
    this.#route(target, frameDataReq(msgId, data));
  }

  #resend(key) {
    const p = this.pending.get(key);
    if (!p) return;
    this.pending.delete(key);
    if (p.retries <= 0) return; // give up after MAX_RETRIES
    if (this.conns.length === 0) this.buffer.push({ target: p.target, data: p.data });
    else this.#transmit(p.target, p.data, p.retries - 1);
  }

  #flushBuffer() {
    const queued = this.buffer;
    this.buffer = [];
    for (const { target, data } of queued) this.#dispatch(target, data);
  }

  #received(sender, e) {
    switch (e.type) {
      case Type.DATA:
        this.inbox.push(new Message(e.payload, sender));
        break;
      case Type.DATA_REQ:
        this.inbox.push(new Message(e.payload, sender));
        this.#route(sender, frameAck(e.msgId)); // ack on the same connection
        break;
      case Type.ACK: {
        const p = this.pending.get(e.msgId.toString("hex"));
        if (p) {
          clearTimeout(p.timer);
          this.pending.delete(e.msgId.toString("hex"));
        }
        break;
      }
      // HELLO / HEARTBEAT after handshake: ignored (they only reset the timeout).
    }
  }
}
